### PR TITLE
Notebook for matching function

### PR DESCRIPTION
**Pull request by bmj1988**
This implements the hierarchical matching function, as well as many of the changes recommended in the slack,

- new notebook matching_algorithm_benchmarks.ipynb. contains timing and valid match benchmarks of all the matching functions used so far which basically explains the evolution from the current main branch to this one
- processed OCR returns as a .json (data/processed_ocr_data) for the largest collection of ballots we have available for testing which will eliminate the need to use the chatgpt API
- Adds a "Test Cross Check with Preprocessed Data" button
- logging for all the timing we track in app (default writes to data/logs/benchmark_logs.log). *.log has been added to the gitignore
- rerouted all file creation for temporary files to the temp_ocr_images folder
- changes "uploaded_file" to "uploaded_ballots" for clarity
- vectorized column generation for Full Name, Full Address, and OCR (a combination of full name and full address)
- adds glob and loguru imports (loguru added to requirements.txt)
 
 Mostly eliminated for ... in range() loops. I've kept indexes where it makes sense to track them (progress bars, the naming convention for the jpg files) but used glob to grab all jpgs in the temp_ocr_images folder, to make the code that iterates through them a little more legible.

### CODE REVIEW

File: data/processed_ocr_data.json
Comment (mowillia) [inline data/processed_ocr_data.json:1]: Where does the data in this file come from? If it is OCR from the voter records, then we cannot add it to this public repository.

File: app/app.py [hunk 1]
```diff
@@ -85,49 +88,69 @@ def extract_signature_info(image_path):
...
-# creating full name column
-voter_records_2023_df['Full Name'] = voter_records_2023_df.apply(lambda x: f"{x['First_Name']} {x['Last_Name']}", axis=1)
-full_name_list = list(voter_records_2023_df['Full Name'])
 
+voter_records_2023_df = pd.read_csv('data/raw_feb_23_city_wide.csv', dtype=str)
 
+# creating full name, full address, and OCR columns
+voter_records_2023_df['Full Name'] = voter_records_2023_df["First_Name"] + ' ' + voter_records_2023_df['Last_Name']
+voter_records_2023_df['Full Address'] =  voter_records_2023_df["Street_Number"] + " " + voter_records_2023_df["Street_Name"] + " " + voter_records_2023_df["Street_Type"] + " " + voter_records_2023_df["Street_Dir_Suffix"]
+voter_records_2023_df['OCR'] = voter_records_2023_df["Full Name"] + ' ' + voter_records_2023_df["Full Address"]
```
Comment (mowillia) [inline app/app.py:153]: The 'OCR' column should be called 'Full Name and Full Address'. Calling it 'OCR' is confusing since this column isn't derived from OCR.

File: app/app.py [hunk 2]
```diff
@@ -85,49 +88,69 @@ def extract_signature_info(image_path):
...
-    sorted_dictionary = dict(sorted(full_name_score_dict.items(), reverse=True, key=lambda item: item[1]))
-
-    # top five key value pairs (indices and scores)
-    indices_scores_list = list(sorted_dictionary.items())[:5]
+##
+# TIERED SEARCH
+##
 
-    return indices_scores_list
+def tiered_search(name, address):
```
Comment (mowillia) [inline app/app.py:100]: The `if-else` conditionals in this function are confusingly written. I think the logic is OK but it is hard to parse. 

I tried to rewrite it to be more readable. Let me know if you think the below function does something different

```
def tiered_search(name, address):

    name_address_combo = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR
    name_address_matches = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address_name, name_address_score, name_address_id = name_address_matches[0]

    # if score is more than 85, return the tuple
    if name_address_score >= 85.0:
        return (name_address_name, name_address_score, name_address_id)
    
    # if score is below 85, do additional processing
    else:

        # computing matches based on name and address; only considers all other wards
        name_address__matches = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__name, name_address__score, name_address__id = name_address__matches[0]

        # computing matches based on name alone; considers full voter records
        full_name__matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
        full_name__name, full_name__score, full_name__id = full_name__matches[0]

        # if score based on name and address is greater than score based on full name, then output name and address info
        if name_address__score >= full_name__score:
            return (name_address__name, name_address__score, name_address__id)

        # if name address score is less than full name, then output full name score
        # but first search for close address in voter records
        else:
            address = voter_records_2023_df.loc[full_name__id, 'Full Address']
            full_name_address = f"{full_name__name} {address}"
            return (full_name_address, full_name__score, full_name__id)
```

File: data/processed_ocr_data.json
Comment (bmj1988) [reply data/processed_ocr_data.json:1]: I used filter-repo to wipe it from the commit history so that it can't be accessed or referenced again, it was generated from the petition signatures. I have a local copy I can send you directly so that it can be used by anyone who has approval, I will add it to the gitignore for my next commit

File: app/app.py (hunk 2 above)
Comment (bmj1988) [reply app/app.py:100]: This version functions the same assuming there is a match, but the old version will keep the record with the highest score between the name_address ward search and the name_address registry search. This only matters if none of the searches produces a valid match, and the Ward search returns a higher match % than the Full name search, which may be statistically insignificant. In any case it definitely won't affect valid match % any, just the highest result return when there are no valid matches. I'll update it to use this version,

Comment (mowillia) [reply app/app.py:100]: Ah I see. I missed that case: It's possible that the ward-based score is less than 85, but that the scores for the full registry are less than the ward-based score in which case we would want to return the ward-based results. I think we can encompass this case by slightly rewriting the function. Here is a new version. Let me know if it misses any other cases from the original

```
def tiered_search(name, address):

    name_address_combo = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR
    name_address_matches1 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address__name1, name_address__score1, name_address__id1 = name_address_matches1[0]

    # if score is more than 85, return the tuple
    if name_address__score1 >= 85.0:
        return (name_address__name1, name_address__score1, name_address__id1)
    
    # if score is below 85, do additional processing
    else:

        # computing matches based on name and address; only considers all other wards
        name_address_matches2 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__name2, name_address__score2, name_address__id2 = name_address_matches2[0]

        # computing matches based on name alone; considers full voter records
        full_name_matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
        full_name__name, full_name__score, full_name__id = full_name_matches[0]

        # if original ward-based score is the highest, then return ward-based results
        if (name_address__score1 >= name_address__score2) and (name_address__score1 >= full_name__score):
            return (name_address__name1, name_address__score1, name_address__id1)

        else:
            # if score based on name and address is greater than score based on full name, then output name and address info
            if name_address__score2 >= full_name__score:
                return (name_address__name2, name_address__score2, name_address__id2)

            # if name address score is less than full name, then output full name score
            # but first search for address in voter records
            else:
                address = voter_records_2023_df.loc[full_name__id, 'Full Address']
                full_name_address = f"{full_name__name} {address}"
                return (full_name_address, full_name__score, full_name__id)
```

Comment (bmj1988) [reply app/app.py:100]: I rewrote this slightly, only so we don't have to traverse the full dataframe twice there's a match found in the Full Name + Address registry search. Let me know what you think about the readability.

    def tiered_search(name, address):
    name_address = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR

    name_address_matches = score_fuzzy_match_slim(name_address, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address__name, name_address__score, name_address__id = name_address_matches[0]

    # If no Valid matches are found in the ward, searches for a match against the entire registry except the ward already checked

    if name_address__score < 85.0:
        name_address_matches2 = score_fuzzy_match_slim(name_address, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__score2 = name_address_matches2[0][1]

        # If the Registry search is higher than the Ward search, reassigns values for score check and Full Name score comparison

        if name_address__score < name_address__score2:
            name_address__name, name_address__score_, name_address__id = name_address_matches2[0]

    # If there is a valid match on either search, returns the record
    if name_address__score >= 85.0:
        return (name_address__name, name_address__score, name_address__id)

    # IF no Valid matches have been found, searches for a match using only the Full Name

    else:
        full_name_matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
        full_name__name, full_name__score, full_name__id = full_name_matches[0]

    # Compare scores of full name + address match to score of Full Name match and take the record with the highest score in the format Tuple(matched_record, score, index)

    # If Name + Address score is greater than the Full Name score, return Name + Address Record
    if name_address__score > full_name__score:
        return (name_address__name, name_address__score, name_address__id)

    # Else return the Full Name record
    else:
        address = voter_records_2023_df.loc[full_name__id, 'Full Address']
        full_name_address = f"{full_name__name} {address}"
        return (full_name_address, full_name__score, full_name__id)

Comment (mowillia) [reply app/app.py:100]: This has the same structure as the original code. I previously found it confusing because it had a series of `if` conditionals which had overlapping conditions and thus could be better served by being combined. In particular the final set of `if-else` conditionals should be a sub case of the `else` condition in the second `if-else` conditional since you only get to that final set if you don't `return` the second computation of `(name_address__name, name_address__score, name_addess__id)`. 

But I think I better understand what you're trying to do. I'm trying to maximize readability to make it easier to improve the code later on. Here is my previous code in a more readable form which also avoids the full voter records check if the second ward check satisfies the threshold. 

```
def tiered_search(name, address):

    name_address_combo = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR
    name_address_matches1 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address__name1, name_address__score1, name_address__id1 = name_address_matches1[0]

    # if score is more than 85, return the tuple
    if name_address__score1 >= 85.0:
        return (name_address__name1, name_address__score1, name_address__id1)
    
    # if score is below 85, do additional processing
    else:

        # computing matches based on name and address; only considers all other wards
        name_address_matches2 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__name2, name_address__score2, name_address__id2 = name_address_matches2[0]

        # if the new voter records score is greater than 85, return tuple
        if name_address__score2 >= 85.0:
            return (name_address__name2, name_address__score2, name_address__id2)
        
        # if score is less than 85, perform full records search based on name
        # and return results with highest score
        else: 
            # computing matches based on name alone; considers full voter records
            full_name_matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
            full_name__name, full_name__score, full_name__id = full_name_matches[0]

            # find max from three scores
            max_indx = np.argmax([name_address__score1, name_address__score2, full_name__score])

            # return records associated with that max
            if max_indx==0:
                return (name_address__name1, name_address__score1, name_address__id1)
            elif max_indx ==1: 
                return (name_address__name2, name_address__score2, name_address__id2)
            else: 
                address = voter_records_2023_df.loc[full_name__id, 'Full Address']
                full_name_address = f"{full_name__name} {address}"
                return (full_name_address, full_name__score, full_name__id)
```

**Comment by mowillia** [comment]
There seem to be a lot of files that are included in this PR which don't need to be (e.g., README and java files). I'm guessing this comes from the force pushes. It might be necessary to start a new branch after pulling most recently from the `main` branch.
